### PR TITLE
Rewrite of the doc generation

### PR DESCRIPTION
**Pull request by ph**
**Motivation**: We have decided to rewrite the documentation generator to add more features
to our documentation and to fix some outstanding bugs.

This PR addresses the following problem:
- Hard to add new features to the code base.
- The code was tied with Logstash.
- No easy way to publish the doc for a single plugin
- The plugin author was not able to test their documentation changes
- The reported errors were hard to understand.
- No easy way to automate it.
- the minimal requirement on base classes.
# Features
## Generating the documentation for a Logstash instance

You can now generate the documentation for a Logstash distribution.

_prerequisite_:
- You need to have installed the plugins inside Logstash; this is necessary to pickup the right file when generating the documentation.
- You need to have the development gems installed. `bin/logstash-plugin --no-verify --development`

If you have all of the above you can run this command:

``` sh
rake docs:generate-plugins
rake doc:generate-plugins[/tmp/the-new-doc] # Depending on the shell you are using you might need to quote the task name, like this rake "doc:generate-plugins[/tmp/new-doc]"

```

This will give you an output similar to this

``` sh
logstash-input-file > SUCCESS
logstash-input-s3 > SUCCESS
logstash-input-kafka > FAIL
[...]
Exceptions: XXXX
```

The generator will try to generate the doc for all the plugins defined in the _Gemfile_ and installed in Logstash, if anything goes wrong it won't stop the generation of the other plugin. The Task will also report any errors with stack traces at the end, if one plugin fails the build,
you can interrupt the process, and it will output the current errors before exiting.

The docs will be generated in `build/docs/`.
## Generating the documentation for all the plugins from the organization

You can now generate the documentation from the master for all the plugin in the _logstash-plugins_ organization.

_prerequisite_
- You need to go in the `tools/logstash-docgen` directory
- You need to have the dependency installed.
  
  To get started you can run the following commands:
  
  ``` sh
  cd tools/logstash-docgen
  bundle install
  ```

You can use the the `bin/logstash-docgen` command to generate any plugin that you want; this executable can generate all the plugins or specific one from their master branch.

Usages:

``` sh
bin/logstash-docgen --all # will generate the doc for all the plugins
bin/logstash-docgen logtash-input-file logstash-input-s3 # generate doc for 2 plugins
```

**See:** `bin/logstash-docgen --help` for complete usage.

**Notes:**
- The nature of theses tasks require a lot of external execution to make sure all the doc are done in isolation, this process can take a long time.
- Some plugins will be skipped, see `logstash-docgen.yml` for details.
- The script will inject itself as a dependency on the plugin.

The doc are generate by default in a `target, where the command is executed.
## Testing the documentation as the plugin author

_prerequisite_
- Declare the `logstash-docgen` as development dependency
- Add `require "logstash/docgen/plugin_doc"` to the `Rakefile`
- Run `bundle install`

After you can have access to a few rake tasks, you can list them with `bundle exec rake -T`

```
bundle exec rake doc:asciidoc # return the raw asciidoc
bundle exec rake doc:html # Transform the asciidoc into html and open your default browser
bundle exec rake doc:html:raw # Give you the raw html
```

Fixes: #5762

### CODE REVIEW

**Comment by ph** [comment]
@suyograo Ready for a run of review, the build are failling because `launchy` the gem I used to launch the browser is `ISC` licensed, I believe this is compatible with APL. If its not the case I can easily remove that feature.

I am working to add tests, but they arent easy, i will push another commit with them.

File: rakelib/docs.rake [hunk 1]
```diff
@@ -1,25 +1,22 @@
...
+
+    require "logstash/docgen/logstash_generator"
 
-  task "generate-index" do
-    list = [ 'inputs', 'outputs', 'filters', 'codecs' ]
-    list.each do |type|
-      cmd = "bin/bundle exec ruby docs/asciidoc_index.rb asciidoc_generated #{type}"
-      system(cmd)
-    end
+    FileUtils.mkdir_p(DOC_DIRECTORY)
```
Comment (suyograo) [inline rakelib/docs.rake]: Can we take a flag to configure the output directory?

File: tools/logstash-docgen/README.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
+**Features**
+
+## Generating the documentation for a Logstash instance
```
Comment (suyograo) [inline tools/logstash-docgen/README.md]: Nice! Thanks for putting up a README

**Comment by suyograo** [comment]
@ph Couple of points as I test this:
* Can we have a single script in tools/docgen, for example -- `tools/docgen/from_installed_plugins.sh` which sets up all the prereqs? `rake bootstrap`, `rake test:install-core` (which install the dev dependencies). This script can be used in CI environment.
* I ran `rake docs:generate-plugins`. Checked beats input, the IDs for the config links are repeated. See `<<plugins-inputs-add_field-add_field>>` instead of `<<plugins-inputs-add_field>>`

File: logstash-core/logstash-core.gemspec [hunk 3]
```diff
@@ -44,6 +44,10 @@ Gem::Specification.new do |gem|
   gem.add_runtime_dependency "rubyzip", "~> 1.1.7"
   gem.add_runtime_dependency "thread_safe", "~> 0.3.5" #(Apache 2.0 license)
 
+  # Required by the documentation tools
+  gem.add_runtime_dependency "asciidoctor", "~> 1.5.4"
+  gem.add_runtime_dependency "launchy", "~> 2.4.3"
```
Comment (suyograo) [inline logstash-core/logstash-core.gemspec]: Hmm, I am a bit iffy about `launchy`. If it's not essential, and the license is not in our list, i would vote to remove it. You can always launch this stuff from your browser manually, right?

File: tools/logstash-docgen/lib/logstash/docgen/parser.rb [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    def initialize(options = {})
+      @config = {}
+      @options = options
+    end
+
+    def default_plugin
+      # HACK: until this is merged https://github.com/elastic/logstash/pull/5096
+      # I have to use some trickery to have the default plugins list.
+      begin
+        path = ::File.expand_path(::File.join(::File.dirname(__FILE__), "..", "..", "..", "..", "..", "rakelib", "default_plugins"))
```
Comment (suyograo) [inline tools/logstash-docgen/lib/logstash/docgen/parser.rb]: wow :)

File: tools/logstash-docgen/lib/logstash/docgen/parser.rb [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+    # Used for making `variables` available inside
+    # ERB templates.
+    def get_binding
+      binding
+    end
+  end
+
+
+  # This class is the main point of entry to the parsing of the plugin ruby file,
```
Comment (suyograo) [inline tools/logstash-docgen/lib/logstash/docgen/parser.rb]: Thanks for all the comments. Really nice.

File: tools/logstash-docgen/logstash-docgen.yml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+---
+ignore_plugins:
```
Comment (suyograo) [inline tools/logstash-docgen/logstash-docgen.yml]: Woohoo. nice

File: logstash-core/logstash-core.gemspec (hunk 3 above)
Comment (ph) [inline logstash-core/logstash-core.gemspec]: Yup I will remove it, I think `rake doc:html | open chrome` might work?;)

File: tools/logstash-docgen/lib/logstash/docgen/parser.rb (hunk 4 above)
Comment (ph) [inline tools/logstash-docgen/lib/logstash/docgen/parser.rb]: Well In an ideal world I would just pull the json and parse it... :(

But #5096 is really close to get merge.

**Comment by ph** [comment]
@suyograo I have pushed the changes for your initial review. thanks!

**Comment by ph** [comment]
@suyograo Also I put the **ci scripts** in the **ci folder**, I think it is better to have all of them together.

**Comment by ph** [comment]
@suyograo I was missing a commit for the double id, the latest commit fixes the problem with the rabbitmq and also add the release date of the gem by querying the rubygems api.

**Comment by ph** [comment]
Fetching the  compatible versions

```
=== collectd

Version: 3.0.2
Released on: 2016-07-14
https://github.com/logstash-plugins/logstash-codec-collectd/blob/master/CHANGELOG.md#302[Changelog]
Compatible: 5.0.0, 2.3.4, 2.3.3, 2.3.2, 2.3.1
```

**Comment by ph** [comment]
Need another round of review.

Before merging we need to do the following:

- create the logstash-docgem and publish it.
- remove the **:path** in the Logstash's Gemfile (if we don't the artifact will be broken).

**Comment by suyograo** [comment]
Will review today

File: rakelib/test.rake [hunk 7]
```diff
@@ -39,6 +39,7 @@ namespace "test" do
 
   desc "run core specs"
   task "core" => ["setup"] do
+    puts $LOAD_PATH
```
Comment (suyograo) [inline rakelib/test.rake]: Was this a left over debug statement?

File: spec/fixtures/dummy.gemspec [hunk 8]
```diff
@@ -0,0 +1,24 @@
+Gem::Specification.new do |s|
+  s.name            = 'logstash-filter-dummy'
+  s.version         = '0.1.1'
+  s.licenses        = ['Apache License (2.0)']
+  s.summary         = "This plugin is only used in the acceptance test"
+  s.description     = "This plugin is only used in the acceptance test"
+  s.authors         = ["Elasticsearch"]
+  s.email           = 'info@elasticsearch.com'
```
Comment (suyograo) [inline spec/fixtures/dummy.gemspec]: info@elastic.co and Elastic for author

File: tools/logstash-docgen/lib/logstash/docgen.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+require "logstash/docgen/version"
+
+module Logstash
```
Comment (suyograo) [inline tools/logstash-docgen/lib/logstash/docgen.rb]: LogStash instead of `Logstash`?

**Comment by suyograo** [comment]
@ph the links are still wrong. 

They are now `plugins-inputs-add_field` instead of `plugins-inputs-beats-add_field`

**Comment by ph** [comment]
@suyograo updated 

See the output.

https://gist.github.com/ph/017684cedac8401a35ed3fc906943304

Also I want this merged: https://github.com/elastic/logstash/pull/5096

**Comment by suyograo** [comment]
Works like a charm. LGTM. Please squash before merging.

@ph Let's not wait for #5096, we can always update once that is merged.

**Comment by wiibaa** [comment]
@ph another curious question: do you plan to make `logstash-docgen` as a `logstash-devutils` dependency or should it really be added in each plugin ?

**Comment by ph** [comment]
@wiibaa It could be a runtime dependency of `logstash-devutils`, but we were thinking about removing devutils.

**Comment by ph** [comment]
Actually if it's a dependency of devutils, it would actually make more sense.

**Comment by ph** [comment]
@suyograo I cannot merged it before #5096,.because I have to release a version of this gem to make logstash use it as a dev dependency. Logstash currently use this gem in the gemfile with the `:path` that point to a local directory. When we create an artifact we dont include the `logstash-docgen` in the package this will make the package unable to run or install new plugins.


And without #5096 the gem will be broken since it will try to use a local path that doesn't exist.

**Comment by jordansissel** [comment]
Ready to merge, @ph?

**Comment by ph** [comment]
beats and jdbc are causing build error as reported by @suyograo https://gist.github.com/suyograo/24f05da8403262021c753ecfaef6b582

**Comment by suyograo** [comment]
Also when I go to `tools/logstash-docgen` I get this error (even after running `bundle install`):

```
suyog@machine:~/ws/elastic/logstash/tools/logstash-docgen (pr/5720)$ bin/logstash-docgen logstash-input-beats
Processing, 1 plugins: beats
beats > FAIL
--------------------------------------------------------------------------------
FAILURE: beats
       	Exception: LogStash::Docgen::Plugin::CommandException message: cmd: bundler install --jobs 8 --quiet --path /tmp/vendor, stdout: Could not find gem 'flores (~> 0.0.6) java' in any of the gem sources listed in
your Gemfile or available on this machine.
, stderr:
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:104:in `run_in_directory'
       		/Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler.rb:237:in `with_clean_env'
       		/Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler.rb:447:in `with_env'
       		/Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler.rb:237:in `with_clean_env'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:99:in `run_in_directory'
       		org/jruby/RubyDir.java:341:in `chdir'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:98:in `run_in_directory'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:94:in `bundle_install'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:52:in `generate'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:175:in `generate'
       		org/jruby/RubyProc.java:281:in `call'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/task_runner.rb:27:in `run'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:174:in `generate'
       		org/jruby/RubyArray.java:1613:in `each'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:173:in `generate'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/runner.rb:27:in `execute'
       		/Users/suyog/.rvm/gems/jruby-1.7.25/gems/clamp-1.0.0/lib/clamp/command.rb:68:in `run'
       		/Users/suyog/.rvm/gems/jruby-1.7.25/gems/clamp-1.0.0/lib/clamp/command.rb:133:in `run'
       		bin/logstash-docgen:6:in `(root)'
```

**Comment by suyograo** [comment]
with Kafka:

```
suyog@machine:~/ws/elastic/logstash/tools/logstash-docgen (pr/5720)$ bin/logstash-docgen logstash-input-kafka
Processing, 1 plugins: kafka
kafka > FAIL
--------------------------------------------------------------------------------
FAILURE: kafka
       	Exception: LogStash::Docgen::Plugin::CommandException message: cmd: bundle exec rake vendor, stdout: , stderr: Bundler::GemNotFound: Could not find rake-11.2.2 in any of the sources
      materialize at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/spec_set.rb:95
             map! at org/jruby/RubyArray.java:2446
      materialize at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/spec_set.rb:88
            specs at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/definition.rb:140
        specs_for at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/definition.rb:185
  requested_specs at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/definition.rb:174
  requested_specs at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/environment.rb:19
            setup at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/runtime.rb:14
            setup at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler.rb:95
           (root) at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/setup.rb:19
          require at org/jruby/RubyKernel.java:1040
           (root) at /Users/suyog/.rvm/rubies/jruby-1.7.25/lib/ruby/shared/rubygems/core_ext/kernel_require.rb:1
```

**Comment by suyograo** [comment]
Ok, found another issue. For ES output -- configs are got from common_config.rb which is not getting reflected in the doc generator.

https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch/common_configs.rb#L141

https://www.elastic.co/guide/en/logstash/master/plugins-outputs-elasticsearch.html#plugins-outputs-elasticsearch-pipeline

**Comment by ph** [comment]
@suyograo for the ES code I think the issue might be because of `mod.config`, the mod, I think it's not taken into consideration. The regexp expect `config` at the beginning.

**Comment by ph** [comment]
I've fixed the issue with the elasticsearch output, when I was gathering source_location I was only taking into consideration module that defined instance methods, I was wrong.

I have also fixed a render issue when we use `type => :array`

**Comment by suyograo** [comment]
LGTM

**Comment by jordansissel** [comment]
Testing this --

```
⓿ localhost(~/projects/logstash/tools/logstash-docgen) pull/5720
% bundle install
Fetching gem metadata from https://rubygems.org/............
Fetching version metadata from https://rubygems.org/...
Fetching dependency metadata from https://rubygems.org/..
Resolving dependencies...
Bundler could not find compatible versions for gem "bundler":
  In Gemfile:
    bundler (~> 1.12)

  Current Bundler version:
    bundler (1.11.2)
This Gemfile requires a different version of Bundler.
Perhaps you need to update Bundler by running `gem install bundler`?

Could not find gem 'bundler (~> 1.12)' in any of the sources
```

**Comment by ph** [comment]
@jordansissel 
I dont remember adding this dependency on a specific bundler version, It might be the generator shipped with `bundler` that automatically add a version requirement.
https://github.com/elastic/logstash/pull/5720/files#diff-b1c3263ad41e99f8f41a04d5b170d576R28

**Comment by ph** [comment]
@jordansissel I just pushed a new commit that remove the constraints, can you give it another shot?

**Comment by jordansissel** [comment]
docgen fails if my ruby is MRI. Possibly expected, but noting this just in case.

**Comment by jordansissel** [comment]
New issue this time. I use JRuby 1.7.24 for my ruby, and do this:

```
# (in git root, toplevel of logstash repo)
% rake bootstrap
% cd tools/logstash-docgen
% bundle install
% bin/logstash-docgen logstash-input-stdin
Processing, 1 plugins: stdin
stdin > FAIL
--------------------------------------------------------------------------------
FAILURE: stdin
        Exception: LogStash::Docgen::Plugin::CommandException message: cmd: bundler install --jobs 8 --quiet --path /tmp/vendor, stdout: Could not find gem 'logstash-codec-plain java' in any of the gem sources listed
in your Gemfile or available on this machine.
, stderr:
                /home/jls/projects/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:104:in `run_in_directory'
                /home/jls/.rvm/gems/jruby-1.7.24/gems/bundler-1.11.2/lib/bundler.rb:219:in `with_clean_env'
                /home/jls/.rvm/gems/jruby-1.7.24/gems/bundler-1.11.2/lib/bundler.rb:199:in `with_original_env'
```

Logstash works, though `bin/logstash -e ""` reads from stdin happily.

**Comment by ph** [comment]
@jordansissel Is you default ruby MRI or jruby? because the `logstash-docgen` tool will use a shellout and I am wondering if its executed in jruby or mri?

**Comment by jordansissel** [comment]
@ph I've tried with both MRI and Jruby as default ruby :\

**Comment by ph** [comment]
Few modifications:

- make sure we can generate the index
-make sure we can generate the doc locally?

**Comment by ph** [comment]
See this Issue https://github.com/elastic/logstash/issues/6086 for the source of the description in the index page.

**Comment by fbaligand** [comment]
Hi guys,
what is missing in this PR so that it can be merged ?

**Comment by karmi** [comment]
Hi @ph, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/logstash/pull/5720.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

**Comment by karmi** [comment]
Hi @ph, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/logstash/pull/5720.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

**Comment by ph** [comment]
@suyograo I've rebased, squashed this PR and removed the logstash-docgen from the gemfile and created 2 ci scripts for the build system. Since this gem is not pushed on rubygems and we dont it ship in the package we have to make sure we don't declare it in the gemfile. The ci scripts will take of that at build time.

Building docs for the plugins installed installed in logstash works without any problems.
**BUT** building plugins from their master branch fails for some, I have to go through all of them and either fix the issue in the plugin itself or in the doc.

Can you give a last look using the `ci/ci_docs.sh` command and I will merge that in master and make other PRs for the other issues.

**Comment by suyograo** [comment]
```
suyog@machine:~/ws/elastic/logstash (pr/5720)$ ci/ci_docs.sh
Invoking bundler install...
Plugin version conflict, aborting
rake aborted!
Bundler::VersionConflict: Bundler could not find compatible versions for gem "octokit":
  In Gemfile:
    logstash-docgen (>= 0) java depends on
      octokit (~> 4.3.0) java

    octokit (= 3.8.0) java
```

@ph this needs new constraints I guess

**Comment by ph** [comment]
constrains updated.

**Comment by ph** [comment]
@suyograo I've fixed the issue with the %PLUGIN% from the common plugin option, we now have a doc post process that will take care of theses values, the branch is also rebased with the current master.

**Comment by suyograo** [comment]
@ph LGTM let's get this puppy merged

**Pull request closed without merging**